### DISCUSSION TITLE
Control the visibility of a story

### DISCUSSION THREAD

**Discussion by Andreasox** [pymupdf/PyMuPDF#3010] [Looking for help] [answered]
Is it possible to define visibility, i.e. set OCG/layer, for a story object?

**Comment by JorjMcKie** [comment]
No, not directly as such, because the output of a story by definition is a sequence of rectangles, on one or more pages.

However, you can use a story in `Page.insert_htmlbox(rect, story)`. Here, the Story is used instead of a text string. This method support the `oc` parameter.

**Comment by JorjMcKie** [comment]
You can also use [story.write_with_links](https://pymupdf.readthedocs.io/en/latest/story-class.html#Story.write_with_links).
This delivers a new PDF document with possibly multiple pages. The page sizes and the filled rectangles on these pages are under control of Python callbacks (functions) that you provide.

Then take the pages of your target PDF and insert those story pages via [show_pdf_page](https://pymupdf.readthedocs.io/en/latest/page.html?highlight=show_pdf_page#Page.show_pdf_page). I am referring to a snippet like this:
```python
src = story.write_with_links(rect_callback, pagefn=page_callback, ...)

doc = fitz.open()
for srcpage in src:
    page = doc.new_page(width=srcpage.rect.width, height=srcpage.rect.height)
    page.show_pdf_page(page.rect, src, srcpage.number, oc=ocg_xref)
...
```
Just a more or less generic example.
You also take the pages of an existing doc above and fill the same (fixed) rectangle on each of them with the `srcpage` contents, etc.

**Comment by JorjMcKie** [comment]
@Andreasox - does the above make sense?

**Comment by Andreasox** [comment]
Yes, thanks a lot. Then I now how we can handle this. 
But I must admit that I do not really understand why an oc tag cannot be directly assigned to a story.

**Reply by JorjMcKie** [reply, marked as answer]
A Story is no PDF-relevant object at all.

In PDF, only certain objects can be subject to OC visibility control: images, Form XObjects, annotations and segments within a page's `/Contents` objects (to cover text and vector graphics).

BTW a PDF page cannot be subject to OC either.

Story obviously is none of the above. More than anything else it is a HTML DOM plus a number of methods to achieve placement on **_rectangles_**. A Story even doesn't even know what a PDF is.
Its output producing methods `.place()` and `.draw()` can be **_directed_** to a new PDF using the DocumentWriter class, at which point, the Story seizes to exist ... in a way.
Only the output to a PDF which it has created in cooperation with a DocumentWriter can possibly be put under control of OC in some of the ways I showed above.
